evm/src/recursive_verifier.rs

```diff
@@ -34,118 +49,615 @@ use crate::{
+    fn from_vec(v: &[T], config: &StarkConfig) -> Self {
+        let mut iter = v.iter().copied();
+        let trace_cap = (0..1 << config.fri_config.cap_height)
+            .map(|_| next_chunk::<_, 4>(&mut iter).to_vec())
+            .collect();
+        let ctl_challenges = GrandProductChallengeSet {
+            challenges: (0..config.num_challenges)
+                .map(|_| GrandProductChallenge {
+                    beta: iter.next().unwrap(),
+                    gamma: iter.next().unwrap(),
+                })
+                .collect(),
+        };
+        let challenger_state_before = next_chunk(&mut iter);
+        let challenger_state_after = next_chunk(&mut iter);
+        let ctl_zs_last = iter.collect();
+
+        Self {
+            trace_cap,
+            ctl_zs_last,
+            ctl_challenges,
+            challenger_state_before,
+            challenger_state_after,
+        }
+    }
+}
+
+impl<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>
+    RecursiveAllProof<F, C, D>
+{
+    /// Verify every recursive proof.
+    pub fn verify(
+        self,
+        verifier_data: &[VerifierCircuitData<F, C, D>; NUM_TABLES],
+        cross_table_lookups: Vec<CrossTableLookup<F>>,
+        inner_config: &StarkConfig,
+    ) -> Result<()>
+    where
+        [(); C::Hasher::HASH_SIZE]:,
+    {
+        let pis: [_; NUM_TABLES] = std::array::from_fn(|i| {
+            PublicInputs::from_vec(&self.recursive_proofs[i].public_inputs, inner_config)
+        });
+
+        let mut challenger = Challenger::<F, C::Hasher>::new();
+        for pi in &pis {
+            for h in &pi.trace_cap {
+                challenger.observe_elements(h);
+            }
+        }
+        let ctl_challenges =
+            get_grand_product_challenge_set(&mut challenger, inner_config.num_challenges);
+        // Check that the correct CTL challenges are used in every proof.
+        for pi in &pis {
+            ensure!(ctl_challenges == pi.ctl_challenges);
+        }
+
+        let state = challenger.compact();
+        ensure!(state == pis[0].challenger_state_before);
+        // Check that the challenger state is consistent between proofs.
+        for i in 1..NUM_TABLES {
+            ensure!(pis[i].challenger_state_before == pis[i - 1].challenger_state_after);
+        }
+
+        // Verify the CTL checks.
+        let degrees_bits = std::array::from_fn(|i| verifier_data[i].common.degree_bits);
+        verify_cross_table_lookups::<F, C, D>(
+            cross_table_lookups,
+            pis.map(|p| p.ctl_zs_last),
+            degrees_bits,
+            ctl_challenges,
+            inner_config,
+        )?;
+
+        // Verify the proofs.
+        for (proof, verifier_data) in self.recursive_proofs.into_iter().zip(verifier_data) {
+            verifier_data.verify(proof)?;
+        }
+        Ok(())
+    }
+
+    /// Recursively verify every recursive proof.
+    pub fn verify_circuit(
+        self,
```

Comment:
I think this should not take `self` (the list of recursive proofs), it seems unused and requiring it would mean we can't build this circuit without doing some proving work.